src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,390 @@
+  is_bound_at_inf (f * g) := by simpa using hf.mul hg
+
+@[simp]lemma bound_mem (f : ℍ → ℂ) :
+  is_bound_at_inf f ↔ ∃ (M A : ℝ), ∀ z : ℍ, A ≤ im z → abs (f z) ≤ M :=
+begin
+  simp [is_bound_at_inf, asymptotics.is_O_iff, filter.eventually, at_I_inf_mem],
+end
+
+/-- A function `f : ℍ → ℂ` is a modular form weight `k ∈ ℤ` and of level `Γ` if it is holomorphic,
+ weakly modular and bounded at infinity -/
+structure is_modular_form_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) (f : ℍ → ℂ) : Prop :=
+  (hol : mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (↑f : ℍ' → ℂ))
+  (transf : f ∈ weakly_modular_submodule k Γ)
+  (infinity : ∀ (A : SL(2,ℤ)), is_bound_at_inf (f ∣[k] A))
+
+/-- A function `f : ℍ → ℂ` is a cusp form of weight `k ∈ ℤ` and of level `Γ` if it is holomorphic,
+ weakly modular, and zero at infinity -/
+structure is_cusp_form_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) (f : ℍ → ℂ) : Prop :=
+  (hol : mdifferentiable 𝓘(ℂ) 𝓘(ℂ) (↑f : ℍ' → ℂ))
+  (transf : f ∈ weakly_modular_submodule k Γ)
+  (infinity : ∀ (A : SL(2,ℤ)), is_zero_at_inf (f ∣[k] A))
+
+/-- The zero modular form is a cusp form-/
+lemma zero_cusp_form : is_cusp_form_of_weight_and_level k Γ 0 :=
+{ hol := by {apply mdifferentiable_zero,},
+  transf := (weakly_modular_submodule k Γ).zero_mem',
+  infinity := by { intro A,
+    convert zero_form_is_zero_at_inf,
+    rw slash,
+    simp only [pi.zero_apply, zero_mul],
+    refl, } }
+
+lemma is_modular_form_of_weight_and_level_of_is_cusp_form_of_weight_and_level {f : ℍ → ℂ}
+  (h : is_cusp_form_of_weight_and_level k Γ f) : is_modular_form_of_weight_and_level k Γ f :=
+{ hol := h.1,
+  transf := h.2,
+  infinity := λ (A : SL(2,ℤ)), is_zero_at_inf_is_bound _ (h.3 A), }
+
+ /-- The zero modular form is a modular form-/
+lemma zero_mod_form : is_modular_form_of_weight_and_level k Γ 0 :=
+begin
+  apply_rules [is_modular_form_of_weight_and_level_of_is_cusp_form_of_weight_and_level,
+    zero_cusp_form],
+end
+
+/-- This is the space of modular forms of weight `k` and level `Γ`-/
+def space_of_mod_forms_of_weight_and_level (k : ℤ) (Γ : subgroup SL(2,ℤ)) : submodule ℂ (ℍ → ℂ) :=
+{ carrier := { f : ℍ → ℂ | is_modular_form_of_weight_and_level k Γ f},
+  zero_mem':= by { simp only [set.mem_set_of_eq], apply zero_mod_form, },
+  add_mem' := by { intros a b ha hb,
+    split,
+    exact mdifferentiable_add ha.hol hb.hol,
+    exact (weakly_modular_submodule k Γ).add_mem' ha.transf hb.transf,
+    intro A, rw slash_add,
+    exact (bounded_at_infty_submodule.add_mem' (ha.infinity A) (hb.infinity A)) },
+  smul_mem' := by { intros c f hf,
+    split,
+    exact mdifferentiable_smul _ hf.hol,
+    exact (weakly_modular_submodule k Γ).smul_mem' _ hf.transf,
+    intro A,
+    rw smul_slash,
+    apply (bounded_at_infty_submodule.smul_mem' c (hf.infinity A)), }, }
+
+localized "notation `M(`k`, `Γ`)`:= space_of_mod_forms_of_weight_and_level k Γ" in modular_forms
```

Comment:
I'm not a big fan of this notation, why not simply `localized "notation M := space_of_mod_forms_of_weight_and_level"`. Adding the brackets back when almost all function calls in lean don't seems counterintuitive.

Reply:
I'm all for trying to make things look like normal maths for non-leaners, but people will always be confronted with a lot of functions applied to arguments without brackets when looking at lean

Reply:
Fair enough. I did want to make it look more normal, but you are right, it doesn't fit with the rest of mathlib,

Reply:
Notice that [continuous_map](https://leanprover-community.github.io/mathlib_docs/topology/continuous_function/basic.html#continuous_map) uses the notation `C(α, β)`, and in this PR there are still notations like `GL(2, ℝ)` and `SL(2, ℤ)` that remain.

Reply:
The SL ones are local notations though, which is more like a personal preference of the author thing as it only shows up when working on the file and not in the docs I believe, so I didn't complain about those ;).
Are the C ones normal brackets or some weird unicode?

Reply:
[Those are normal parentheses and `C` isn't even localized](https://github.com/leanprover-community/mathlib/blob/bd6b98b22cdb9f13a7e1fd790a8e796d723f8c0f/src/topology/continuous_function/basic.lean#L32).
```
import topology.continuous_function.basic
example {α} [topological_space α] (f : C(α,α)) : f = f := rfl
```
simply works.

Reply:
Wow! I can't help but think this must cause issues occasionally, but i guess they can't be too bad